app/uk/gov/hmrc/exports/services/reversemapping/declaration/DeclarationXmlParser.scala

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.services.reversemapping.declaration
+
+import uk.gov.hmrc.exports.services.reversemapping.declaration.DeclarationXmlParser.XmlParserResult
+
+import scala.xml.NodeSeq
+
+trait DeclarationXmlParser[A] {
+  def parse(inputXml: NodeSeq): XmlParserResult[A]
+}
+
+object DeclarationXmlParser {
+  type XmlParserResult[A] = Either[XmlParsingException, A]
+
+  implicit class EitherList[T](list: Seq[XmlParserResult[T]]) {
+    def toEitherOfList: XmlParserResult[Seq[T]] = {
+      def insideOut(rest: Seq[XmlParserResult[T]], acc: Seq[T]): XmlParserResult[Seq[T]] = rest match {
```

Comment:
Nit: could make this a tail-recursive method

Reply:
According to IJ it is. I know that the self-invocation is not in the last `case` statement, but perhaps the compiler does the work?

Reply:
pretty sure you need the @tailrec annotation in scala to make compiler turn this into a tail-recursive function (and yeah, last call needs to be the insideOut call so maybe need to re-order case statement).

Reply:
@tailrec is just for compile-time checking if the method is indeed tail-recursive. And it seems it does not need to be the last case statement.